src/sqlfluff/dialects/dialect_postgres.py

```diff
@@ -1420,13 +1420,15 @@ class CreateTableStatementSegment(ansi.CreateTableStatementSegment):
                             Sequence(
                                 Ref("ColumnReferenceSegment"),
                                 Ref("DatatypeSegment"),
-                                Sequence(
-                                    "COLLATE",
-                                    Ref("QuotedLiteralSegment"),
-                                    optional=True,
-                                ),
                                 AnyNumberOf(
-                                    Ref("ColumnConstraintSegment", optional=True)
+                                    OneOf(
+                                        Ref("ColumnConstraintSegment", optional=True),
+                                        Sequence(
```

Comment:
If I read this correctly, this allows any number of `COLLATE`s, e.g.:
```
COLLATE a COLLATE b COLLATE c
```

Is this the intent? I don't see any tests doing this. Should we add one?

Reply:
From reading the docs, I think only **one** `COLLATE` clause is permitted.

https://www.postgresql.org/docs/current/collation.html

Reply:
Related to the test I put, just ran this in Postgres and it worked:

```
gregoryfinley=# CREATE TABLE orders3
(
order_number text NOT NULL COLLATE numeric,
order_number2 text COLLATE numeric NOT NULL
);
CREATE TABLE
gregoryfinley=# 
```

Reply:
That makes sense, and it should be okay. The dialect is often more permissive/flexible than what is actually allowed (or makes sense to do).

I'd like @tunetheweb to review this PR as well, since he is more the expert on dialect changes. He's on vacation right now, so I'm doing a preliminary review.

Reply:
Thanks for contributing!

It's funny, 1-2 years ago, we naively thought we'd actually finish with the dialects someday. There is a never-ending supply of missing things, even with the popular dialects like Postgres, Snowflake, and BigQuery. Incredible. 🤯

Reply:
Yep! I have considered crazy ideas like, could we build a parser to auto-convert Postgres (etc.) docs into sqlfluff match grammar. 🤔

Reply:
I've also wondered, at least for the commercial data stores, could we get vendors to do the work.

Reply:
Or sponsor the project 🤞 .

Reply:
`AnyNumberOf` supports a `max_times_per_element`, which if set to 1 is the same as `AnySetOf` so can use that instead

Reply:
This works:

```py
                                AnySetOf(
                                    Sequence(
                                        "COLLATE",
                                        Ref("ObjectReferenceSegment"),
                                        optional=True,
                                    ),
                                    AnyNumberOf(
                                        Ref("ColumnConstraintSegment"),
                                        optional=True,
                                    ),
                                ),
```

Note it won't allow `ColumnConstraintSegment` + `COLLATE clause` + `ColumnConstraintSegment` if that's valid?

Reply:
You can do a `ColumnConstraintSegment` before and after:

```
gregoryfinley=# CREATE TABLE orders4
(
order_number2 text PRIMARY KEY COLLATE numeric NOT NULL
);
CREATE TABLE
gregoryfinley=# 
```

Maybe what we have currently is fine?

Reply:
Yeah I’m happy to merge as is if you are. Won’t be the first place the syntax is slightly overly permissive and usually that’s better than overly restrictive.

Reply:
Yeah I'm ok with it. I pushed another commit with more tests on multiple column constraints: https://github.com/sqlfluff/sqlfluff/pull/3571/commits/e5ad52363566221e50d7c701fc80dd05810245fb

Reply:
You actually removed two
of the tests and replaced it with one all-inclusive test. I wonder if we should keep those two tests in though as well, in case anyone removes the all-inclusive test as an unlikely, unnecessary complication but doesn’t then realise order is unimportant too?

Reply:
I think I did the opposite, moving one CREATE to two? I did just to use PRIMARY KEY in different places in the syntax.

Reply:
I think this works better: https://github.com/sqlfluff/sqlfluff/pull/3571/commits/010cb8b298781d9158bb7d125ecdf1f88956d8ac

Reply:
Ok but now we don’t have the `CONSTRAINT` + `COLLATE` + `CONSTRAINT` example so someone could try to clean this up in future using the `AnySetOf` syntax I proposed above and not realise we specifically choose NOT to use that. So I think you need all three tests for complete coverage?

Though, to be honest, think it’s pretty unlikely anyone would actually use that third type, but there’s all sorts out there!!

Reply:
I think my test _is_ an example of `CONSTRAINT` + `COLLATE` + `CONSTRAINT`, isn't it?

`order_number text UNIQUE COLLATE numeric NOT NULL`

`UNIQUE` is a constraint, then `COLLATE`, then `NOT NULL` is a second constraint.

Or am I missing something?

Reply:
Ah you’re right. What you’re missing then is `COLLATE` + `CONSTRAINT` example.

Though really getting picky now so happy enough to just merge if you prefer?

Reply:
This should be everything? Let me know if not: https://github.com/sqlfluff/sqlfluff/pull/3571/commits/0f95ac5bb961755638ff2f9dfcdffeedbb804818